brayns/engine/model/Components.h

```diff
@@ -0,0 +1,158 @@
+/* Copyright (c) 2015-2022, EPFL/Blue Brain Project
+ * All rights reserved. Do not distribute without permission.
+ * Responsible Author: Nadir Roman Guerrero <nadir.romanguerrero@epfl.ch>
+ *
+ * This file is part of Brayns <https://github.com/BlueBrain/Brayns>
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License version 3.0 as published
+ * by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#pragma once
+
+#include <algorithm>
+#include <cassert>
+#include <memory>
+#include <typeindex>
+#include <vector>
+
+namespace brayns
+{
+class Components
+{
+public:
+    template<typename T, typename... Args>
+    T &add(Args &&...args)
+    {
+        static_assert(std::is_same_v<std::decay_t<T>, T>, "Only decayed components allowed");
+
+        auto &entry = _components.emplace_back(typeid(T), std::make_unique<Component<T>>());
+        auto &component = static_cast<Component<T> &>(*entry.component);
+        auto &data = component.data;
+
+        if constexpr (std::is_aggregate_v<T>)
+        {
+            data = T{std::forward<Args>(args)...};
+        }
+        else
+        {
+            data = T(std::forward<Args>(args)...);
+        }
+
+        return data;
+    }
+
+    template<typename T>
+    T &getOrAdd()
+    {
+        if (auto component = find<T>())
+        {
+            return *component;
+        }
+        return add<T>();
+    }
+
+    template<typename T>
+    T &get()
+    {
+        auto component = find<T>();
+        assert(component);
+        return *component;
+    }
+
+    template<typename T>
+    const T &get() const
+    {
+        auto component = find<T>();
+        assert(component);
+        return *component;
+    }
+
+    template<typename T>
+    T *find()
+    {
+        auto it = _findIterator<T>();
+        if (it == _components.end())
+        {
+            return nullptr;
+        }
+        auto &entry = *it;
+        auto &component = entry.component;
+        auto &cast = static_cast<Component<T> &>(*component);
+        return &cast.data;
+    }
+
+    template<typename T>
+    const T *find() const
+    {
+        auto it = _findIterator<T>();
+        if (it == _components.end())
+        {
+            return nullptr;
+        }
+        auto &entry = *it;
+        auto &component = entry.component;
+        auto &cast = static_cast<Component<T> &>(*component);
+        return &cast.data;
```

Comment:
As the component is a pointer, it can be retreived as a non-const in both find versions (const and non-const) so we could do this to avoid duplications:

T *find() {return _find() }

const T *find() const {return _find() }

T *_find() const { /*block of code above*/}

Reply:
Addressed